src/wp-includes/class-wp-theme-json-resolver.php

```diff
@@ -0,0 +1,360 @@
+
+			// phpcs:ignore WordPress.WP.I18n.LowLevelTranslationFunction,WordPress.WP.I18n.NonSingularStringLiteralText,WordPress.WP.I18n.NonSingularStringLiteralContext,WordPress.WP.I18n.NonSingularStringLiteralDomain
+			$array_to_translate[ $item_key ][ $key ] = translate_with_gettext_context( $array_to_translate[ $item_key ][ $key ], $context, $domain );
+		}
+
+		return $array_to_translate;
+	}
+
+	/**
+	 * Given a theme.json structure modifies it in place
+	 * to update certain values by its translated strings
+	 * according to the language set by the user.
+	 *
+	 * @param array  $theme_json The theme.json to translate.
+	 * @param string $domain    Optional. Text domain. Unique identifier for retrieving translated strings.
+	 *                          Default 'default'.
+	 *
+	 * @return array Returns the modified $theme_json_structure.
+	 */
+	private static function translate( $theme_json, $domain = 'default' ) {
+		$fields = self::get_fields_to_translate();
+		foreach ( $fields as $field ) {
+			$path    = $field['path'];
+			$key     = $field['key'];
+			$context = $field['context'];
+
+			/*
+			 * We need to process the paths that include '*' separately.
+			 * One example of such a path would be:
+			 * [ 'settings', 'blocks', '*', 'color', 'palette' ]
+			 */
+			$nodes_to_iterate = array_keys( $path, '*', true );
+			if ( ! empty( $nodes_to_iterate ) ) {
+				/*
+				 * At the moment, we only need to support one '*' in the path, so take it directly.
+				 * - base will be [ 'settings', 'blocks' ]
+				 * - data will be [ 'color', 'palette' ]
+				 */
+				$base_path = array_slice( $path, 0, $nodes_to_iterate[0] );
+				$data_path = array_slice( $path, $nodes_to_iterate[0] + 1 );
+				$base_tree = _wp_array_get( $theme_json, $base_path, array() );
+				foreach ( $base_tree as $node_name => $node_data ) {
+					$array_to_translate = _wp_array_get( $node_data, $data_path, null );
+					if ( is_null( $array_to_translate ) ) {
+						continue;
+					}
+
+					// Whole path will be [ 'settings', 'blocks', 'core/paragraph', 'color', 'palette' ].
+					$whole_path       = array_merge( $base_path, array( $node_name ), $data_path );
+					$translated_array = self::translate_theme_json_chunk( $array_to_translate, $key, $context, $domain );
+					_wp_array_set( $theme_json, $whole_path, $translated_array );
+				}
+			} else {
+				$array_to_translate = _wp_array_get( $theme_json, $path, null );
+				if ( is_null( $array_to_translate ) ) {
+					continue;
+				}
+
+				$translated_array = self::translate_theme_json_chunk( $array_to_translate, $key, $context, $domain );
+				_wp_array_set( $theme_json, $path, $translated_array );
+			}
+		}
+
+		return $theme_json;
+	}
+
+	/**
+	 * Return core's origin config.
+	 *
+	 * @return WP_Theme_JSON Entity that holds core data.
+	 */
+	public static function get_core_data() {
+		if ( null !== self::$core ) {
+			return self::$core;
+		}
+
+		$config     = self::read_json_file( __DIR__ . '/theme.json' );
```

Comment:
This file only holds data in the WordPress install. However, from the plugin, we'll add more properties (settings, styles) that we want to have a default value as if they were defined by core.

There're several alternatives here. From the point of view of the extenders of this API, it'll be great to have a single point in which we can both extend the allowed settings (see `theme_json_allowed_schema` filter) and provide their default values at once.

Reply:
I've gone with the straightforward path here: add a new filter for data `theme_json_core_data` separated from `theme_json_allowed_schema`.

The rationale is the same I shared https://github.com/WordPress/wordpress-develop/pull/1262#discussion_r636772667 The `theme_json_allowed_schema` filter benefits from being agnostic about the `theme.json` shape: it consumers want to expand the allowed settings they don't have to traverse or know the theme.json shape (top-level vs blocks, etc).

See the filters in action in the plugin: [schema](https://github.com/WordPress/gutenberg/pull/32059/files#diff-0b8e98dead5542e95159bd3c482158111d3f16637e7eef9af13373358fa42dedR298), [data](https://github.com/WordPress/gutenberg/pull/32059/files#diff-0b8e98dead5542e95159bd3c482158111d3f16637e7eef9af13373358fa42dedR266).

Reply:
Why can't we just use the existing `block_editor_settings` filters to add what we want to add in the Gutenberg plugin? Adding these extra filters sounds very premature to me.

Reply:
We don't need them, and I can remove them if we deem them premature. The rationale to introduce these was that, from the plugin, we wouldn't need to do anything else for the post editor to work (we'd still need to add new settings for the site editor and the REST Request, though).

Reply:
Removed them as per https://github.com/WordPress/wordpress-develop/pull/1262/commits/eb50ac508cf16439f05183ab0227a17ce6eb0e73